File: code_challenge_solution/java/120 Mastermind.java
```java
import java.util.*;

public class Mastermind {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Random random = new Random();
        
        int[] secretCode = generateSecretCode(random);
        int[] guess = new int[4];
        
        boolean gameOver = false;
        int attempts = 0;
        
        System.out.println("Welcome to Mastermind!");
        System.out.println("Guess the 4-digit secret code.");
        System.out.println("Each digit must be between 1 and 6.");
        
        while (!gameOver) {
            System.out.print("Enter your guess: ");
            String guessStr = scanner.nextLine();
            
            if (guessStr.length() != 4) {
                System.out.println("Invalid guess. Please enter 4 digits.");
                continue;
            }
            
            for (int i = 0; i < 4; i++) {
                try {
                    guess[i] = Integer.parseInt(guessStr.charAt(i) + "");
                } catch (NumberFormatException e) {
                    System.out.println("Invalid guess. Please enter digits only.");
                    continue;
                }
                
                if (guess[i] < 1 || guess[i] > 6) {
                    System.out.println("Invalid guess. Each digit must be between 1 and 6.");
                    continue;
                }
            }
            
            int[] feedback = getFeedback(secretCode, guess);
            attempts++;
            
            System.out.println("Attempts: " + attempts);
            System.out.println("Feedback: " + Arrays.toString(feedback));
            
            if (Arrays.equals(feedback, new int[] {4, 0})) {
                gameOver = true;
                System.out.println("Congratulations! You have guessed the secret code.");
            }
        }
        
        System.out.println("Game over.");
    }
    
    public static int[] generateSecretCode(Random random) {
        int[] secretCode = new int[4];
        for (int i = 0; i < 4; i++) {
            secretCode[i] = random.nextInt(6) + 1;
        }
        return secretCode;
    }
    
    public static int[] getFeedback(int[] secretCode, int[] guess) {
        int[] feedback = new int[2];
        int[] secretCodeCount = new int[6];
        int[] guessCount = new int[6];
        
        for (int i = 0; i < 4; i++) {
            if (secretCode[i] == guess[i]) {
                feedback[0]++;
            } else {
                secretCodeCount[secretCode[i] - 1]++;
                guessCount[guess[i] - 1]++;
            }
        }
        
        for (int i = 0; i < 6; i++) {
            feedback[1] += Math.min(secretCodeCount[i], guessCount[i]);
        }
        
        return feedback;
    }
}
```

File: code_challenge_solution/c#/120 Mastermind.cs
using System;

namespace Mastermind
{
    public class Program
    {
        public static void Main(string[] args)
        {
            Console.WriteLine("Welcome to Mastermind!");

            Game game = new Game();
            game.Start();

            Console.WriteLine("Thanks for playing!");
        }
    }

    public class Game
    {
        private const int MaxTries = 10;
        private int currentTry;
        private string secretCode;

        public void Start()
        {
            currentTry = 1;
            GenerateSecretCode();

            Console.WriteLine("Guess the 4-digit code (each digit should be between 1 and 6)");

            while (currentTry <= MaxTries)
            {
                string guess = GetUserGuess();
                string feedback = GetFeedback(guess);

                Console.WriteLine(feedback);

                if (guess == secretCode)
                {
                    Console.WriteLine("Congratulations! You cracked the code!");
                    return;
                }

                currentTry++;
            }

            Console.WriteLine("Game over. You ran out of tries. The secret code was: " + secretCode);
        }

        private void GenerateSecretCode()
        {
            Random random = new Random();
            secretCode = "";

            for (int i = 0; i < 4; i++)
            {
                secretCode += random.Next(1, 7);
            }
        }

        private string GetUserGuess()
        {
            Console.Write("Enter your guess: ");
            return Console.ReadLine();
        }

        private string GetFeedback(string guess)
        {
            int correctPosition = 0;
            int correctColor = 0;

            for (int i = 0; i < 4; i++)
            {
                if (guess[i] == secretCode[i])
                {
                    correctPosition++;
                }
                else if (secretCode.Contains(guess[i].ToString()))
                {
                    correctColor++;
                }
            }

            return "Correct position: " + correctPosition + ", Correct color: " + correctColor;
        }
    }
}